hazelcast/src/main/java/com/hazelcast/nio/tcp/WriteHandler.java

```diff
@@ -172,7 +172,7 @@ private void unschedule() {
         // So the outputBuffer is empty, so we are going to unschedule ourselves.
         scheduled.set(false);
 
-        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty()) {
+        if (writeQueue.isEmpty() && urgentWriteQueue.isEmpty() && currentPacket == null) {
```

Comment:
@jerrinot, of course!

@sancar, @pveentjer; 
I find the new WriteHandler impl more readable and understandable. Old one was working good but it was hard to understand the code. Java NIO API is already very complex, as Peter said. Actually it's more complex than it should be, I think. 

I personally don't mind minor bugs which can be easily fixed in a patch release. Since we are testing this change since 3.4, it seems working. This one, for instance, is a critical one, which occurs when system is under very low load and causes big latency jump. It's hidden from our eyes thanks to heartbeat messages sent periodically :). 

 @jerrinot: can you also add a test to show the issue and verify the problem? You can increase the heartbeat interval to a very large value in test to prevent it to interleave...